src/context/index.ts

```diff
@@ -0,0 +1,187 @@
+		store.logs.removeWhere(log => log._uri === doc.uri.toString())
+	})
+
+	// Actions/Decorations for Call Trees
+	const decorationTypeCallout = window.createTextEditorDecorationType({
+		after: { color: new ThemeColor('problemsWarningIcon.foreground') }
+	})
+	const decorationTypeHighlight = window.createTextEditorDecorationType({
+		border: '1px',
+		borderStyle: 'solid',
+		borderColor: new ThemeColor('problemsWarningIcon.foreground'),
+	})
+	languages.registerCodeActionsProvider('*', {
+		provideCodeActions: (doc, _range, context) => {
+			if (context.only) return
+
+			const diagnostic = context.diagnostics[0] as ResultDiagnostic | undefined
+			if (!diagnostic) return
+
+			const result = diagnostic?.result
+			panel.select(result)
+
+			const editor = window.visibleTextEditors.find(editor => editor.document === doc)
+			if (!editor) return // When would editor be undef?
+			if (!result) return // Don't clear the decorations until the next result is selected.
+
+			const locations = result?.codeFlows?.[0]?.threadFlows?.[0]?.locations ?? []
+			const messages = locations.map((tfl, i) => {
+				const text = tfl.location?.message?.text
+				return `Step ${i + 1}${text ? `: ${text}` : ''}`
+			})
+			const ranges = locations.map(tfl => regionToSelection(doc, parseRegion(tfl.location?.physicalLocation?.region)))
+			const rangesEnd = ranges.map(range => {
+				const endPos = doc.lineAt(range.end.line).range.end
+				return new Range(endPos, endPos)
+			})
+			const rangesEndAdj = rangesEnd.map(range => {
+				const tabCount = doc.lineAt(range.end.line).text.match(/\t/g)?.length ?? 0
+				const tabCharAdj = tabCount * (editor.options.tabSize as number - 1) // Intra-character tabs are counted wrong.
+				return range.end.character + tabCharAdj
+			})
+			const maxRangeEnd = Math.max(...rangesEndAdj) + 2 // + for Padding
+			const decorCallouts = rangesEnd.map((range, i) => ({
+				range,
+				hoverMessage: messages[i],
+				renderOptions: { after: { contentText: ` ${'┄'.repeat(maxRangeEnd - rangesEndAdj[i])} ${messages[i]}`, } }, // ←
+			}))
+			editor.setDecorations(decorationTypeCallout, decorCallouts)
+			editor.setDecorations(decorationTypeHighlight, ranges)
+			return []
+		}
+	})
+	observe(store.logs, change => {
+		const {removed} = change as unknown as IArraySplice<Log>
+		if (!removed.length) return
+		window.visibleTextEditors.forEach(editor => {
+			editor.setDecorations(decorationTypeCallout, [])
+			editor.setDecorations(decorationTypeHighlight, [])
+		})
+	})
+
+	// Virtual Documents
+	workspace.registerTextDocumentContentProvider('sarif', {
+		provideTextDocumentContent: (uri, token) => {
+			const [logUriEncoded, runIndex, artifactIndex] = uri.path.split('/')
+			const logUri = decodeURIComponent(logUriEncoded)
+			const artifact = store.logs.find(log => log._uri === logUri)?.runs[+runIndex]?.artifacts?.[+artifactIndex]
+			const contents = artifact?.contents
+			if (contents?.text) return contents?.text
```

Comment:
>contents?.text [](start = 7, length = 14)

This can have markdown as well.

Reply:
Is this high priority for you guys?

---
In reply to: [435428710](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r435428710) [](ancestors = 435428710)